### ISSUE TITLE
Support Oracle?

### ISSUE THREAD

**Issue by ted9219** [OdyOSG/CDMConnector#18]
Hi, I'm just wondering if the CDMConnector package supports the Oracle database.

I'm quite not familiar with this package yet (I usually use the DatabaseConnector), but I need to connect with a database using this package for collaboration with European folks. 

It seems that the ODBC can make us connect with Oracle DB but I couldn't see examples for Oracle connections in the README file. 
Is there anyone who has experience with Oracle Database?

**Comment by ablack3** [comment]
Hi @ted9219, thanks for trying this package!
Yes Oracle is supported. I'd recommend using odbc but I'm open to supporting connections made with DatabaseConnector as they are also DBI connections. However DatabaseConnector connections are not fully tested so you might have issues. 

My recommendation is to first set up odbc driver for Oracle on your system. This can be a painful process but once it is done you shouldn't have to change it.

See: https://www.oracle.com/uk/database/technologies/releasenote-odbc-ic.html

Once you confirm you can connect using odbc try running this reprex.

```r
library(CDMConnector)

"OracleODBC-19" %in% odbc::odbcListDataSources()$name
#> [1] TRUE

con <- DBI::dbConnect(odbc::odbc(), "OracleODBC-19")

writeSchema <- "OHDSI"
cdmSchema <- "CDMV5"


listTables(con, schema = cdmSchema)
#>  [1] "ACHILLES_ANALYSIS"        "ACHILLES_HEEL_RESULTS"   
#>  [3] "ACHILLES_RESULTS"         "ACHILLES_RESULTS_DERIVED"
#>  [5] "ACHILLES_RESULTS_DIST"    "ATTRIBUTE_DEFINITION"    
#>  [7] "CARE_SITE"                "CDM_SOURCE"              
#>  [9] "COHORT"                   "COHORT_ATTRIBUTE"        
#> [11] "COHORT_DEFINITION"        "CONCEPT"                 
#> [13] "CONCEPT_ANCESTOR"         "CONCEPT_CLASS"           
#> [15] "CONCEPT_RELATIONSHIP"     "CONCEPT_SYNONYM"         
#> [17] "CONDITION_ERA"            "CONDITION_OCCURRENCE"    
#> [19] "COST"                     "DEATH"                   
#> [21] "DEVICE_EXPOSURE"          "DOMAIN"                  
#> [23] "DOSE_ERA"                 "DRUG_ERA"                
#> [25] "DRUG_EXPOSURE"            "DRUG_STRENGTH"           
#> [27] "FACT_RELATIONSHIP"        "LOCATION"                
#> [29] "MEASUREMENT"              "NOTE"                    
#> [31] "OBSERVATION"              "OBSERVATION_PERIOD"      
#> [33] "PAYER_PLAN_PERIOD"        "PERSON"                  
#> [35] "PROCEDURE_OCCURRENCE"     "PROVIDER"                
#> [37] "RELATIONSHIP"             "SOURCE_TO_CONCEPT_MAP"   
#> [39] "SPECIMEN"                 "VISIT_DETAIL"            
#> [41] "VISIT_OCCURRENCE"         "VOCABULARY"

cdm <- cdm_from_con(con, cdm_schema = cdmSchema, cdm_tables = tbl_group("default"))

verify_write_access(con, write_schema = writeSchema)
#> Error in verify_write_access(con, write_schema = writeSchema): could not find function "verify_write_access"

names(cdm)
#>  [1] "person"               "observation_period"   "visit_occurrence"    
#>  [4] "condition_occurrence" "drug_exposure"        "procedure_occurrence"
#>  [7] "measurement"          "observation"          "death"               
#> [10] "location"             "care_site"            "provider"            
#> [13] "drug_era"             "dose_era"             "condition_era"       
#> [16] "cdm_source"           "concept"              "vocabulary"          
#> [19] "concept_relationship" "concept_synonym"      "concept_ancestor"    
#> [22] "drug_strength"

head(cdm$concept)
#> # Source:   SQL [6 x 10]
#> # Database: Oracle 19.00.0000[OHDSI@orasqltest.cqnqzwtn5s1q.us-east-1.rds.amazonaws.com/ORCL/]
#>   concept_id concept_name               domain_id vocabulary_id concept_class_id
#>        <dbl> <chr>                      <chr>     <chr>         <chr>           
#> 1   45956741 Not applicable             Observat… SNOMED        Pharma/Biol Pro…
#> 2   45956742 Intrauterine device        Observat… SNOMED        Pharma/Biol Pro…
#> 3   45956743 Epidural infusion          Observat… SNOMED        Pharma/Biol Pro…
#> 4   45956744 Powder and solvent for so… Observat… SNOMED        Pharma/Biol Pro…
#> 5   45956745 Powder and solvent for su… Observat… SNOMED        Pharma/Biol Pro…
#> 6   45956746 Powder and solvent for di… Observat… SNOMED        Pharma/Biol Pro…
#> # ℹ 5 more variables: standard_concept <chr>, concept_code <chr>,
#> #   valid_start_date <dttm>, valid_end_date <dttm>, invalid_reason <chr>

dbms(cdm)
#> [1] "oracle"

dplyr::inner_join(cdm$person,
                        cdm$observation_period,
                        by = "person_id") %>%
  head(2) %>%
  dplyr::collect()
#> # A tibble: 2 × 23
#>   person_id gender_concept_id year_of_birth month_of_birth day_of_birth
#>       <dbl>             <dbl>         <dbl>          <dbl>        <dbl>
#> 1       540              8507          1936              2            1
#> 2       600              8507          1935             11            1
#> # ℹ 18 more variables: time_of_birth <chr>, race_concept_id <dbl>,
#> #   ethnicity_concept_id <dbl>, location_id <dbl>, provider_id <dbl>,
#> #   care_site_id <dbl>, person_source_value <chr>, gender_source_value <chr>,
#> #   gender_source_concept_id <dbl>, race_source_value <chr>,
#> #   race_source_concept_id <dbl>, ethnicity_source_value <chr>,
#> #   ethnicity_source_concept_id <dbl>, birth_datetime <dttm>,
#> #   observation_period_id <dbl>, observation_period_start_date <dttm>, …

DBI::dbDisconnect(con)
```

**Comment by ted9219** [comment]
Thanks, @ablack3 it seems very helpful. I will try this in the environment of the data partner site and will keep in touch with you if there are issues.

**Comment by ablack3** [comment]
Hi @ted9219, I'm going to close this issue. But feel free to open another issue if a specific problem comes up. You can also find me in OHDSI Teams if you want to chat about something.

**Comment by ted9219** [comment]
Hi, Just for sharing my experience with the Oracle database. After the installation/configuration of Oracle Instant Client Basic and Instant Client ODBC, it works very well. Thank you.
